server/routerlicious/packages/services-telemetry/src/lumber.ts

```diff
@@ -0,0 +1,149 @@
+// ILumberjackEngine to emit the data according to the engine implementation.
+// Lumber should be created through Lumberjack. Additional properties can be added through
+// addProperty(). Once the telemetry event is complete, the user must call either success()
+// or error() on Lumber to emit the data.
+export class Lumber<T extends string = LumberEventName> {
+    private readonly _eventName: T;
+    private readonly _properties: Map<string, any>;
+    private readonly _timestamp: string;
+    private readonly startTime: number;
+    private readonly _type: LumberType;
+    private _metadata: ITelemetryMetadata | undefined;
+    private _latencyInMs: number | undefined;
+    private _successful: boolean | undefined;
+    private _message: string | undefined;
+    private _statusCode: string | undefined;
+    private _exception: Error | undefined;
+    private _logLevel: LogLevel | undefined;
+    private completed: boolean;
+    private readonly engineList: ILumberjackEngine[];
+
+    public get eventName(): T {
+        return this._eventName;
+    }
+
+    public get metadata(): ITelemetryMetadata | undefined {
+        return this._metadata;
+    }
+
+    public get properties(): Map<string, any> {
+        return this._properties;
+    }
+
+    public get type(): LumberType {
+        return this._type;
+    }
+
+    public get timestamp(): string {
+        return this._timestamp;
+    }
+
+    public get latencyInMs(): number | undefined {
+        if (this._type === LumberType.Log) {
+            return undefined;
+        }
+        return this._latencyInMs;
+    }
+
+    public get successful(): boolean | undefined {
+        if (this._type === LumberType.Log) {
+            return undefined;
+        }
+        return this._successful;
+    }
+
+    public get message(): string | undefined {
+        return this._message;
+    }
+
+    public get statusCode(): string | undefined {
+        return this._statusCode;
+    }
+
+    public get exception(): Error | undefined {
+        return this._exception;
+    }
+
+    public get logLevel(): LogLevel | undefined {
+        return this._logLevel;
+    }
+
+    constructor(
+        eventName: T,
+        type: LumberType,
+        engineList: ILumberjackEngine[]) {
+        this._eventName = eventName;
+        this._type = type;
+        this.completed = false;
+        this._properties = new Map<string, any>();
+        this.startTime = Date.now();
+        this._timestamp = new Date(this.startTime).toISOString();
+        this.engineList = engineList;
+        this._latencyInMs = undefined;
+        this._successful = undefined;
+        this._message = undefined;
+        this._statusCode = undefined;
+        this._exception = undefined;
+        this._logLevel = undefined;
+    }
+
+    public addProperty(key: string, value: any): this {
+        this._properties.set(key, value);
+        return this;
+    }
+
+    public success(
```

Comment:
Is a Lumber's success or error state mapped to the log level of the log being emitted? If so, do we need the explicit public success and error methods on the lumber? The caller that utilizes these 2 methods will likely check the log's level and exception status, and so could they just call the lumber class's emit method directly?

Reply:
Success and Error are associated with Lumber's representation of a metric. For metrics, success or error are tracked in a property called `successful`, present in Lumber.

In this PR, I'm only adding the classes without putting them in use; but in follow-up PRs, it will be possible to see that when we want to capture an event or metric, having explicit `success` or `failure events might make more sense when instrumenting the code. Personally, I think that they make it easier to understand what is happening to Lumber when instrumenting the code; in a try catch block, we call `success` after the operation completes with no errors, and `error` otherwise. If we used `emit`, we would have the same `emit` keyword in both try and catch blocks.

When it comes to the `LogLevel`, it is associated with Lumber for 2 reasons: (i) Lumber is also used internally by Lumberjack to capture log data; (ii) even though Lumber traditionally represents a metric, it also provides specific information more suitable to logs, such as documentId. `LogLevel` is automatically taken care of by Lumberjack if calling Lumberjack's `log` method. And it is only needed for metrics if the user, for some reason, wants to explicitly override the default `LogLevel` values associated with `success` (info) and `error` (error).